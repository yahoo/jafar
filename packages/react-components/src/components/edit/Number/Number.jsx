/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import PropTypes from 'prop-types';
import Input from '@material-ui/core/Input';

/**
 * Represent a number
 * 
 * Import <a target="_blank" 
 href="https://github.com/yahoo/jafar/blob/master/packages/react-components/src/components/edit/Number/Number.jsx">
 Number</a> from '@jafar/react-components/edit/Number'
 */
export default class Number extends React.Component {
  static propTypes = {
    value: PropTypes.number,
    state: PropTypes.shape({
      placeholder: PropTypes.string,
      min: PropTypes.number,
      max: PropTypes.number,
      step: PropTypes.number,
      endAdornment: PropTypes.string,
      startAdornment: PropTypes.string,
    }),
    disabled: PropTypes.bool,
    invalid: PropTypes.bool,
    onValueChange: PropTypes.func.isRequired,
  };

  static defaultProps = {
    value: undefined,
    state: {
      placeholder: '',
      min: undefined,
      max: undefined,
      step: undefined,
      endAdornment: undefined,
      startAdornment: undefined,
    },
    disabled: false,
    invalid: false,
  };

  render() {
    const value = this.props.value === undefined ? '' : this.props.value; // Fixing Material controlled VS uncontrolled issue
    const { state } = this.props;
    const endAdornment = state.endAdornment ? <div style={{ marginLeft: '5px' }}>{state.endAdornment}</div> : undefined;
    const startAdornment = state.startAdornment ? <div style={{ marginRight: '5px' }}>{state.startAdornment}</div> : undefined;

    return (
      <Input
        type="number"
        value={value}
        placeholder={state.placeholder}
        disabled={this.props.disabled}
        error={this.props.invalid}
        onChange={this.onValueChange}
        inputProps={{
          min: state.min,
          max: state.max,
          step: state.step,
        }}
        endAdornment={endAdornment}
        startAdornment={startAdornment}
      />
    );
  }

  onValueChange = (e) => {
    let value = parseFloat(e.target.value);
    value = isNaN(value) ? undefined : value; // eslint-disable-line
    this.props.onValueChange(value);
  }
}
