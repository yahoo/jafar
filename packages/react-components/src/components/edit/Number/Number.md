<h3>Examples</h3>

<h4>Usage in jafar form</h4>

```javascript
const Form = require('@jafar-org/react-form/Form').default;
const Field = require('@jafar-org/react-form/Field').default;

const model = {
  id: 'simple',
  fields: {
    age: {
      label: 'Age',
      path: 'age',
      component: {
        name: 'Number',
        state: {
            placeholder: 'Enter Age',
            min: 0,
        }
      },
      required: true,
      disableTerm: {
          name: 'disableMe',
      },
      excludeTerm: {
          name: 'excludeMe',
      },
    },
  },
  data: {
    age: 18, // remove to see initial value undefined, and required error
  }
};

const resources = {
  components: { 
    Number: {
      renderer: Number,
    },
  },
  terms: {
      disableMe: { func: () => false }, // convert to true to see the field disabled
      excludeMe: { func: () => false }, // convert to true to see the field excluded
  },
};

<Form model={model} resources={resources}>
    <Field id='age' />
</Form>
```

<h4>Simple</h4>

```javascript
initialState = { 
    value: 8,
    state: {
        placeholder: 'Enter Number',
        min: 0,
        max: 10,
        step: 1,
        endAdornment: '$',
    },
};

<div>
 <Number
    value={state.value}
    state={state.state}
    onValueChange={(value) => {
        setState({ value });
    }}
    />
    </div>
```

<h4>Initial value undefined</h4>

```javascript
initialState = { 
    value: undefined,
    state: {
        placeholder: 'Enter Number',
    },
};
 <Number
    value={state.value}
    state={state.state}
    onValueChange={(value) => {
        setState({ value });
    }}
    />
```

<h4>Disabled</h4>

```javascript
initialState = { 
    value: 42,
    state: {
        placeholder: 'Enter Number',
    },
};
 <Number
    value={state.value}
    state={state.state}
    disabled={true}
    onValueChange={(value) => {
        setState({ value });
    }}
    />
```

<h4>Invalid</h4>

```javascript
initialState = { 
    value: 42,
    state: {
        placeholder: 'Enter Number',
    },
};
 <Number
    value={state.value}
    state={state.state}
    invalid={true}
    onValueChange={(value) => {
        setState({ value });
    }}
    />
```