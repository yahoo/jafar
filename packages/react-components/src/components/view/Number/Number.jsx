/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import PropTypes from 'prop-types';

/**
 * Represent a number value
 * 
 * Import <a target="_blank" 
 href="https://github.com/yahoo/jafar/blob/master/packages/react-components/src/components/view/Number/Number.jsx">
 Number</a> from '@jafar/react-components/view/Number'
 */
export default class Number extends React.Component {
  static propTypes = {
    value: PropTypes.number,
    state: PropTypes.shape({
      fixed: PropTypes.number,
      template: PropTypes.string,
    }),
  };

  static defaultProps = {
    value: undefined,
    state: {
      fixed: undefined,
      template: undefined,
    },
  };

  render() {
    const state = this.props.state;
    let value = this.props.value;

    if (value !== undefined && value !== null) {
      if (state.fixed) {
        value = value.toFixed(state.fixed);
      }
      value = value.toString().replace(/\d(?=(\d{3})+\.)/g, '$&,');
      if (state.template) {
        value = state.template.replace('$NUMBER', value);
      }
    }
    
    return (<div>{value}</div>);
  }
}
