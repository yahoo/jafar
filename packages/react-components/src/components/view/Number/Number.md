<h3>Examples</h3>

<h4>Usage in jafar form</h4>

```javascript
const Form = require('@jafar/react-form/Form').default;
const Field = require('@jafar/react-form/Field').default;

const model = {
  id: 'simple',
  fields: {
    vacationDays: {
      label: 'Vacation Days',
      path: 'vacationDays',
      component: {
        name: 'Number',
      }
    },
  },
  data: {
      vacationDays: 12,
  },
};

const resources = {
  components: { 
    Number: {
      renderer: Number,
    },
  }
};

<Form model={model} resources={resources}>
    <Field id='vacationDays' />
</Form>
```

<h4>Simple</h4>

```javascript
initialState = { 
    value: 14556.678,
    state: { 
      fixed: 2,
      template: '$NUMBER $',
    },
};

<Number value={state.value} state={state.state} />
```
