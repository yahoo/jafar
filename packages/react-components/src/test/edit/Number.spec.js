/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import { mount, shallow } from 'enzyme';
import Number from '../../edit/Number/index';

describe('<Number />', () => {
  let component;
  let onValueChangeSpy;
  const value = 42;
  const state = { placeholder: 'Enter Number' };
  const disabled = false;
  const invalid = false;

  beforeEach(() => {
    onValueChangeSpy = jest.fn();
  });

  it('should render provided data', () => {
    component = shallow(
      getComponent(value, state, disabled, invalid, onValueChangeSpy)
    );
    expect(component).toMatchSnapshot();
  });

  it('should trigger onValueChange callback on value change', () => {
    component = mount(
      getComponent(value, state, disabled, invalid, onValueChangeSpy)
    );
    const newValue = 43;
    component.find('input').simulate('change', { target: { value: newValue } });
    expect(onValueChangeSpy).toHaveBeenCalledWith(newValue);
  });

  function getComponent(value, state, disabled, invalid, onValueChange) {
    return (
      <Number
        value={value}
        state={state}
        disabled={disabled}
        invalid={invalid}
        onValueChange={onValueChange}
      />
    );
  }
});
