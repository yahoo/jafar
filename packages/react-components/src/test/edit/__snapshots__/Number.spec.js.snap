// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Number /> should render provided data 1`] = `
<WithStyles(ForwardRef(Input))
  disabled={false}
  error={false}
  inputProps={
    Object {
      "max": undefined,
      "min": undefined,
      "step": undefined,
    }
  }
  onChange={[Function]}
  placeholder="Enter Number"
  type="number"
  value={42}
/>
`;
